graphqlmetrics/server.go

```diff
@@ -0,0 +1,97 @@
+package graphqlmetrics
+
+import (
+	"context"
+	"errors"
+	"github.com/wundergraph/cosmo/graphqlmetrics/gen/proto/wg/cosmo/graphqlmetrics/v1/graphqlmetricsv1connect"
+	"go.uber.org/zap"
+	brotli "go.withmatt.com/connect-brotli"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+	"net/http"
+	"time"
+)
+
+// Option defines the method to customize Server.
+type Option func(s *Server)
+
+type Server struct {
+	server     *http.Server
+	listenAddr string
+	logger     *zap.Logger
+
+	metricsService graphqlmetricsv1connect.GraphQLMetricsServiceHandler
+}
+
+func NewServer(metricsService graphqlmetricsv1connect.GraphQLMetricsServiceHandler, opts ...Option) *Server {
+	s := &Server{
+		metricsService: metricsService,
+		listenAddr:     ":4005",
+		logger:         zap.NewNop(),
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	s.bootstrap()
+
+	return s
+}
+
+func (s *Server) bootstrap() {
+	mux := http.NewServeMux()
+	path, handler := graphqlmetricsv1connect.NewGraphQLMetricsServiceHandler(
+		s.metricsService,
+		// Compress responses with Brotli.
+		brotli.WithCompression(),
+	)
+	mux.Handle(path, handler)
+
+	s.server = &http.Server{
+		Addr: s.listenAddr,
+		// https://ieftimov.com/posts/make-resilient-golang-net-http-servers-using-timeouts-deadlines-context-cancellation/
+		ReadTimeout:       1 * time.Minute,
+		WriteTimeout:      2 * time.Minute,
+		ReadHeaderTimeout: 20 * time.Second,
+		// Use h2c so we can serve HTTP/2 without TLS.
+		Handler:  h2c.NewHandler(mux, &http2.Server{}),
```

Comment:
Is this a good setup for prod as well?

Reply:
Good question. I have no experience in regard to buf connect. It is supported in our infrastructure. We can also use HTTP1.1

Reply:
Using HTTP1.1 for now.